api/xds_protocol.rst

```diff
@@ -651,6 +651,36 @@ adding/removing/updating clusters. On the other hand, routes are not
 warmed, i.e., the management plane must ensure that clusters referenced
 by a route are in place, before pushing the updates for a route.
 
+.. _xds_protocol_TTL:
+
+TTL
+~~~
+
+In the event that the management server becomes unreachable, the last known configuration received
+by Envoy will persist until the connection is reestablished. For some services, this may not be
+desirable. For example, in the case of a fault injection service, a management server crash at the
+wrong time may leave Envoy in an undesirable state. The TTL setting allows Envoy to remove a set of
+resources after a specified period of time if contact with the management server is lost. This can
+be used, for example, to terminate a fault injection test when the management server can no longer
+be reached.
+
+For clients that support the *envoy.config.supports-resource-ttl* client feature, A TTL field may
+be specified on each :ref:`Resource <envoy_api_msg_Resource>`. Each resource will have its own TTL
+expiry time, at which point the resource will be expired. Each xDS type may have different ways of
+handling such an expiry.
+
+To update the TTL associated with a *Resource*, the management server resends the resource with a
```

Comment:
+1

Reply:
If it's not too hard, I'd suggest to do it now, just to make sure we don't accidentally release a version that has this change without that one.  If there are Envoys in the wild that have this change but not that one, management servers won't know whether they can omit the resource when updating a TTL.

Reply:
That's a good point, I'll take a look. I ran into some issues with it when doing it for SotW with the top level TTL implementation, but I think having it all be per resource might make that easier to implement.

Reply:
Alright so it turns out that one xDS implementation already encodes meaning into the `resource` missing: https://github.com/envoyproxy/envoy/blob/master/source/common/router/vhds.cc#L76-L80

This makes it tricky to now overload a missing resource to mean a TTL heartbeat: either we prevent these heartbeat updates from being propagated to subscribers (this might break VHDS? Or at least require some changes) or we update all subscriptions to handle these empty resources. 

I think ideally VHDS should not encode this meaning into an empty resource (seems like it should use a wrapper message within the resource), but that's a somewhat invasive change.

Another less than ideal change would be to include a new parameter that indicates that a resource is a heartbeat resource vs an empty resource. The increase in the API complexity seems undesirable here

Reply:
That seems like a very odd behavior for VHDS.  It uses the incremental protocol only.  Shouldn't the server instead populate [resources_removed](https://github.com/envoyproxy/envoy/blob/1ca53123c078b6545d393d2974c58ddc3bebabb4/api/envoy/api/v2/discovery.proto#L214) for newly requested resources that don't exist?

Ugh... doing a little digging, it looks like I was actually the one who suggested this behavior (see https://github.com/envoyproxy/envoy/pull/7415#discussion_r353853404) *sheepish grin*.  But I suspect the only reason I was thinking in that direction was the weird way that VHDS currently uses aliases.  @htuch, this is probably something we should try to clean up as part of adding the udpa resource names.

In any case, I guess for now we're stuck with the current behavior.

Thanks for looking into it, Snow.

Reply:
I think maybe just add a boolean indicating that this is a pure TTL refresh.

Reply:
This TTL PR is going to land before VHDS moves across to udpa:// naming.

Reply:
Yeah, I realize that.  So I think for now, we should just require the resource to be re-sent.

Reply:
Can we limit this to VHDS?

Reply:
That could work too -- we could basically say that VHDS is the only one that has different behavior, and we can document the fact that we intend to remove that special behavior in the future as part of eliminating aliases.  We'd just have to make sure to structure the APIs in a way that makes it obvious that future code changes should not introduce special behavior for any other resource type -- e.g., maybe we can call the method something like `onResourceExpireVhdsOnlyDoNotUse()`. :)

Reply:
Actually, nevermind that last sentence -- I was conflating this with the other discussion about how to handle expiry on a per-resource basis.  But this is actually only about refreshing the TTLs.

Anyway, it still seems fine to me to put in a temporary special case for VHDS and make everything else behave the right way.  I'll leave it up to the two of you to figure out how to actually implement that.